jars/src/org/toast/activegates/AddGateData.kt

```diff
@@ -0,0 +1,37 @@
+package org.toast.activegates
+
+import com.fs.starfarer.api.campaign.InteractionDialogAPI
+import com.fs.starfarer.api.campaign.rules.MemoryAPI
+import com.fs.starfarer.api.impl.campaign.rulecmd.BaseCommandPlugin
+import com.fs.starfarer.api.util.Misc
+
+class AddGateData : BaseCommandPlugin() {
+    override fun execute(ruleId: String, dialog: InteractionDialogAPI?,
```

Comment:
since i know nothing about kotlin, what is the significance of the question mark?

Reply:
Marks a type as nullable. Kotlin will enforce this both at compile time and runtime.

```kotlin
val nullableStr: String? = null // fine
val nonNullStr: String = null // compiler exception!
```

Since java doesn't have nullable types, when interopting with java you kinda just tell kotlin what to expect...and if you're wrong (eg in the example above, if we received a `null` `ruleId` from some java code), it'll explode at runtime.

Overall, a really nice system and one of the best arguments to use kotlin.